### ISSUE TITLE
Allow to disable empty item option for SelectInput

### ISSUE THREAD

**Issue by weissu42** [marmelab/react-admin#7632]
**Is your feature request related to a problem? Please describe.**
I use the SelectInput for a required property, thus it makes no sense to allow to choose the empty option.

I checked the docs and finally also the code and while it is possible to customize the emptyText and emptyValue, it seems there is no way to *not* render it at all ([SelectInput.tsx](https://github.com/marmelab/react-admin/blob/master/packages/ra-ui-materialui/src/input/SelectInput.tsx#L287)). It looks like this should be pretty easy to add (however I am of course by no means familiar with the code).

**Describe the solution you'd like**
Add an option to disable rendering the empty option, something like:
`<SelectInput source="foo" choices={fooChoices} showEmptyOption={false} />`
Of course this can simply default to `true` to ensure it's not a breaking change

**Describe alternatives you've considered**
* I can add `validate={[required('Foo cannot be empty')]}` to at least prevent the user from saving the form without picking a value, but it would be much nicer to not give users the opportunity to empty the input in the first place

**Comment by djhi** [comment]
That's because react-hook-form requires a default value to be set, which is an empty string. Adding this option would mean to check that when it the `showEmptyOption` is `false`, developers also provided a valid default value. I don't think we should as it make the `SelectInput` API more complicated both for us and for developpers.

**Comment by ghost91-** [comment]
To be honest I find the current behavior pretty confusing. It's completely unexpected that I get an additional entry in the select, in addition to the ones I specify as `choices`. The fact that there currently is no way to remove that empty additional choice just makes it even worse.

In order to not make it more complicated than necessary, you could simply use the first element of the choices as default, if no default is given explicitly. That wouldn't make the `SelectInput` API any more complicated.

**Comment by soullivaneuh** [comment]
I agree with @weissu42 about the non-sens part of having the possibility to select an empty value when the form is required.

What we add before 4.x:

<details>

![image](https://user-images.githubusercontent.com/1698357/169255057-58505727-39ba-4740-94ef-a4ce27e902c6.png)

</details>

What we have now:

<details>

![image](https://user-images.githubusercontent.com/1698357/169255093-59ed8151-6244-4814-b90a-d43d645fb188.png)

</details>

> Adding this option would mean to check that when the showEmptyOption prop is false, developers also provided a valid default value. I don't think we should as it make the SelectInput API more complicated both for us and for developpers.

@djhi This may be doable with conditional typing. However, I don't think we should require a default value when we don't want an empty option. For our use case, we have to ensure the value of the select is selected by a human to avoid  mistakes because of the default value.

In a nutshell, the same behavior as for 3.x.

**Comment by ralfarama** [comment]
I agree with both @ghost91- and @soullivaneuh that there are many situations where having a spurious blank value makes no sense.  There are many scenarios and common ones where one must pick from a set of options and a non-choice is not an acceptable option and the added empty entry (especially as the default) makes for a confusing interface which creates erroneous data by default.  Adding a blank entry is inappropriate, confusing and problematic for many scenarios.  

Furthermore, requiring the developer to set parameters like "EmptyText" and "EmptyValue" to override a blank default is a solution which introduces problems of its own.  When one has a field that requires a value from a referenced column one an implementer is forced to choose between hardcoding a foreign key value or dealing with the case where a user selects an unacceptable blank value.  Either case foists problems on an implementer to deal with an undesired artifact of react-admin's behavior.

**Comment by fzaninotto** [comment]
If the field isn't required, the empty option is necessary to empty it. 
If the field is required, the empty option will always lead to an invalid form. 

We'll go with removing the empty option if the field is required.

**Comment by fzaninotto** [comment]
Fixed by #8039
